File: pbhhg_py/builtins/arithmetics.py
import functools
import math
import operator
from typing import Sequence

from pbhhg_py import abstract_syntax as AS
from pbhhg_py import error
from pbhhg_py import utils


def build_tbl(
    proc_functional: utils.ProcFunctional,
) -> dict[str, AS.Evaluation]:
    del proc_functional  # Unused

    def _all(
        metadata: AS.Metadata, argv: Sequence[AS.Value]
    ) -> AS.EvalContext:
        # short-circuiting
        for arg in argv:
            [arg] = yield from utils.match_arguments(
                metadata, [arg], AS.Boolean
            )
            if not arg.value:
                return AS.Boolean(False)
        return AS.Boolean(True)

    def _multiply(
        metadata: AS.Metadata, argv: Sequence[AS.Value]
    ) -> AS.EvalContext:
        utils.check_min_arity(metadata, argv, 1)
        first = yield argv[0]
        [first] = utils.check_type(metadata, [first], AS.Number | AS.Boolean)
        if isinstance(first, AS.Boolean):
            return (yield from _all(metadata, argv))

        rest = yield from utils.match_arguments(metadata, argv[1:], AS.Number)
        values = (arg.value for arg in [first, *rest])
        return utils.guessed_wrap(
            functools.reduce(operator.mul, values)
        )  # No init value

    def _any(
        metadata: AS.Metadata, argv: Sequence[AS.Value]
    ) -> AS.EvalContext:
        # short-circuiting
        for arg in argv:
            [arg] = yield from utils.match_arguments(
                metadata, [arg], AS.Boolean
            )
            if arg.value:
                return AS.Boolean(True)
        return AS.Boolean(False)

    def _add(
        metadata: AS.Metadata, argv: Sequence[AS.Value]
    ) -> AS.EvalContext:
        utils.check_min_arity(metadata, argv, 1)
        first = yield argv[0]
        [first] = utils.check_type(
            metadata, [first], AS.Number | AS.Boolean | AS.Sequence | AS.Dict
        )
        if isinstance(first, AS.Boolean):
            return (yield from _any(metadata, argv))

        argv = yield from utils.map_strict(argv)
        if isinstance(first, AS.List):
            argv = utils.check_type(metadata, argv, AS.List)
            return AS.List(tuple(item for seq in argv for item in seq.value))
        if isinstance(first, AS.String):
            argv = utils.check_type(metadata, argv, AS.String)
            return AS.String("".join(a.value for a in argv))
        if isinstance(first, AS.Bytes):
            argv = utils.check_type(metadata, argv, AS.Bytes)
            return AS.Bytes(b"".join(a.value for a in argv))
        if isinstance(first, AS.Dict):
            argv = utils.check_type(metadata, argv, AS.Dict)
            merged = {h: (k, v) for a in argv for k, h, v in a.table}
            table = [(k, h, v) for h, (k, v) in merged.items()]
            return AS.Dict(table)
        argv = utils.check_type(metadata, argv, AS.Number)
        return utils.guessed_wrap(sum(a.value for a in argv))

    def _exponentiate(
        metadata: AS.Metadata, argv: Sequence[AS.Value]
    ) -> AS.EvalContext:
        argv = yield from utils.match_arguments(
            metadata, argv, AS.Number, [2, 3]
        )
        if len(argv) == 2:
            base, exponent = argv
            try:
                return utils.guessed_wrap(base.value**exponent.value)
            except ZeroDivisionError:
                raise error.UnsuspectedHangeulDivisionError(
                    metadata, "0의 역수를 구하려고 했습니다."
                ) from None

        argv = utils.check_type(metadata, argv, AS.Integer)
        base, exponent, modulo = [a.value for a in argv]
        try:
            return utils.guessed_wrap(pow(base, exponent, abs(modulo)))
        except ValueError:
            raise error.UnsuspectedHangeulArithmeticError(
                metadata, f"법 {modulo}에 대한 {base}의 {exponent}제곱을 구할 수 없습니다."
            ) from None

    def _integer_division(
        metadata: AS.Metadata, argv: Sequence[AS.Value]
    ) -> AS.EvalContext:
        argv = yield from utils.match_arguments(metadata, argv, AS.Real, 2)
        dividend, divider = [arg.value for arg in argv]
        try:
            value = dividend // divider
            if value < 0:
                value = -(-dividend // divider)
        except ZeroDivisionError:
            raise error.UnsuspectedHangeulDivisionError(
                metadata, "0으로 나누려고 했습니다."
            ) from None
        return utils.guessed_wrap(value)

    def _remainder(
        metadata: AS.Metadata, argv: Sequence[AS.Value]
    ) -> AS.EvalContext:
        argv = yield from utils.match_arguments(metadata, argv, AS.Real, 2)
        if utils.is_type(argv, AS.Integer):
            dividend, divider = [arg.value for arg in argv]
            try:
                if dividend // divider >= 0:
                    return AS.Integer(dividend % divider)
                else:
                    return AS.Integer(-(-dividend % divider))
            except ZeroDivisionError:
                raise error.UnsuspectedHangeulDivisionError(
                    metadata, "0으로 나누려고 했습니다."
                ) from None
        dividend, divider = [arg.value for arg in argv]
        return AS.Float(math.fmod(dividend, divider))

    return {
        "ㄱ": _multiply,  # 곱셈
        "ㄷ": _add,  # 덧셈
        "ㅅ": _exponentiate,  # 거듭제곱
        "ㄴㄴ": _integer_division,  # 몫
        "ㄴㅁ": _remainder,  # 나머지
    }

File: pbhhg_js/src/builtins/arithmetics.ts
import Complex from 'complex.js'

import * as AS from '../abstractSyntax'
import * as E from '../error'
import { add, div, mod, mul, pow } from '../numbers'
import {
  checkArity,
  checkMinArity,
  checkType,
  extractValue,
  isType,
  joinArrayBuffer,
} from '../utils'

export function wrapNumber(num: bigint | number | Complex) {
  if (typeof num === 'bigint') return new AS.IntegerV(num)
  if (num instanceof Complex) return new AS.ComplexV(num)
  return new AS.FloatV(num)
}

/** Returns (g, x, y) such that a*x + b*y = g = gcd(a, b) */
function extendedGcd(a: bigint, b: bigint): [bigint, bigint, bigint] {
  let [x, x_old, y, y_old] = [0n, 1n, 1n, 0n]
  while (a !== 0n) {
    const quotient = b / a
    ;[a, b] = [b % a, a]
    ;[y, y_old] = [y_old, y - quotient * y_old]
    ;[x, x_old] = [x_old, x - quotient * x_old]
  }
  return [b, x, y]
}

function modularInverse(num: bigint, modulo: bigint) {
  const [g, inverse, _] = extendedGcd(num % modulo, modulo)
  if (g === 1n) return (inverse + modulo) % modulo
  throw Error(`법 ${modulo}에 대한 ${num}의 역원이 없습니다.`)
}

function modPow(base: bigint, exponent: bigint, modulo: bigint): bigint {
  if (modulo === 0n) throw Error(`0을 법으로 할 수 없습니다.`)
  base = base % modulo
  let result = 1n
  if (exponent < 0n) {
    exponent = -exponent
    base = modularInverse(base, modulo)
  }
  while (exponent > 0n) {
    if (base === 0n) return 0n
    if (exponent % 2n === 1n) result = (result * base) % modulo
    exponent /= 2n
    base = base ** 2n % modulo
  }
  return result
}

function _multiply(
  context: AS.EvalContextBase,
  metadata: AS.Metadata,
  argv: AS.Value[]
) {
  checkMinArity(metadata, argv, 1)
  const first = checkType(
    metadata,
    [context.strict(argv[0])],
    [AS.BooleanV, ...AS.NumberV]
  )
  if (isType(first, AS.NumberV)) {
    const _argv = checkType(
      metadata,
      argv.map((x) => context.strict(x)),
      AS.NumberV
    )
    return wrapNumber(_argv.map((arg) => arg.value).reduce(mul))
  }

  // short-circuiting
  for (const arg of argv) {
    const [_arg] = checkType(metadata, [context.strict(arg)], [AS.BooleanV])
    if (!_arg.value) return new AS.BooleanV(false)
  }
  return new AS.BooleanV(true)
}

function _add(
  context: AS.EvalContextBase,
  metadata: AS.Metadata,
  argv: AS.Value[]
) {
  checkMinArity(metadata, argv, 1)
  const first = checkType(
    metadata,
    [context.strict(argv[0])],
    [AS.BooleanV, AS.DictV, ...AS.NumberV, ...AS.SequenceV]
  )
  if (isType(first, [AS.BooleanV])) {
    // short-circuiting
    for (const arg of argv) {
      const [_arg] = checkType(metadata, [context.strict(arg)], [AS.BooleanV])
      if (_arg.value) return new AS.BooleanV(true)
    }
    return new AS.BooleanV(false)
  }

  if (isType(first, [AS.StringV])) {
    const _argv = checkType(
      metadata,
      argv.map((x) => context.strict(x)),
      [AS.StringV]
    )
    return new AS.StringV(_argv.map((arg) => arg.str).join(''))
  }
  if (isType(first, [AS.BytesV])) {
    const _argv = checkType(
      metadata,
      argv.map((x) => context.strict(x)),
      [AS.BytesV]
    )
    return new AS.BytesV(joinArrayBuffer(_argv.map(extractValue)))
  }
  if (isType(first, [AS.ListV])) {
    const _argv = checkType(
      metadata,
      argv.map((x) => context.strict(x)),
      [AS.ListV]
    )
    return new AS.ListV(_argv.map(extractValue).reduce((a, b) => a.concat(b)))
  }
  if (isType(first, [AS.DictV])) {
    const _argv = checkType(
      metadata,
      argv.map((x) => context.strict(x)),
      [AS.DictV]
    )
    const result = {}
    _argv.map(extractValue).forEach((d) => Object.assign(result, d))
    return new AS.DictV(result)
  }
  const _argv = checkType(
    metadata,
    argv.map((x) => context.strict(x)),
    AS.NumberV
  )
  return wrapNumber(_argv.map((arg) => arg.value).reduce(add))
}

function _exponentiate(
  context: AS.EvalContextBase,
  metadata: AS.Metadata,
  argv: AS.Value[]
) {
  checkArity(metadata, argv, [2, 3])
  const _argv = checkType(
    metadata,
    argv.map((x) => context.strict(x)),
    AS.NumberV
  )
  if (_argv.length === 3) {
    const [base, exp, mod] = checkType(metadata, _argv, [AS.IntegerV])
    try {
      return new AS.IntegerV(modPow(base.value, exp.value, mod.value))
    } catch (err) {
      if (err instanceof Error) {
        throw new E.UnsuspectedHangeulArithmeticError(metadata, err.message)
      }
      throw err
    }
  }
  checkArity(metadata, _argv, 2)
  try {
    return wrapNumber(pow(_argv[0].value, _argv[1].value))
  } catch (error) {
    if (error instanceof Error) {
      throw new E.UnsuspectedHangeulDivisionError(metadata, error.message)
    }
    throw error
  }
}

function _integerDivision(
  context: AS.EvalContextBase,
  metadata: AS.Metadata,
  argv: AS.Value[]
) {
  checkArity(metadata, argv, 2)
  const _argv = checkType(
    metadata,
    argv.map((x) => context.strict(x)),
    AS.RealV
  )
  try {
    return wrapNumber(div(_argv[0].value, _argv[1].value))
  } catch (error) {
    if (error instanceof Error) {
      throw new E.UnsuspectedHangeulDivisionError(
        metadata,
        '0의 역수를 구하려고 했습니다.'
      )
    }
    throw error
  }
}

function _remainder(
  context: AS.EvalContextBase,
  metadata: AS.Metadata,
  argv: AS.Value[]
) {
  checkArity(metadata, argv, 2)
  const _argv = checkType(
    metadata,
    argv.map((x) => context.strict(x)),
    AS.RealV
  )
  try {
    return wrapNumber(mod(_argv[0].value, _argv[1].value))
  } catch (error) {
    if (error instanceof Error) {
      throw new E.UnsuspectedHangeulDivisionError(
        metadata,
        '0의 역수를 구하려고 했습니다.'
      )
    }
    throw error
  }
}

export default {
  ㄱ: _multiply,
  ㄷ: _add,
  ㅅ: _exponentiate,
  ㄴㄴ: _integerDivision,
  ㄴㅁ: _remainder,
}
